packages/flutter/lib/src/painting/image_cache.dart

```diff
@@ -114,6 +113,7 @@ class ImageCache {
   ///   * [ImageProvider], for providing images to the [Image] widget.
   bool evict(Object key) {
```

Comment:
Two questions:

- Does the returned value still make sense if I evict an image that has failed to load? Currently, evict will return false and the image will not be removed from pending images. With this change, it will be removed at still return false.

- Removing an image from pending won't prevent it from adding itself to the cache later if it eventually completes. Should this also remove the cache listener to prevent this from happening, or is it a desirable behavior?

Reply:
* I think the behavior of the evict method should be as follows:
  When the image is in the cache the behavior should be as it is now.
  When the image is not in the cache but in the pending images it should be removed from the pending images and `true` should be returned.
  The cache size should only be adjusted when the image is in the cache (because pending images do not have an effect on the cache size).
  I think to return `true` when the image is still pending is completely fine because the user of the cache does maybe not know if the image is already loaded or not, he only wants it to be removed from it.
  
* In my opinion it would not be correct when the image can add itself to the cache after it is evicted from it. This could lead to unexpected behavior.
  So yes, removing the listener would probably be the right choice.
  For that the inner function `listener` must be moved to class-level. But if it would be private I think there should not be any implications.
  
If you agree I can make the changes.

Reply:
Minor addition: It's not possible with moving the `listener` method to class-level. I solved it with another approach: I introduced a _PendingImage class that holds the completer and a reference to the `listener `method (used for later removal in the evict-case).

Reply:
That sounds like a reasonable approach